internal/controller/virtualmachinefilerestore_controller.go

```diff
@@ -1447,5 +1514,216 @@ func (r *VirtualMachineFileRestoreReconciler) ensureRestoreNamespace(
+	backupToPVCUIDs := make(map[string][]string)
+	backupToSourceNamespace := make(map[string]string) // Track source namespace where PVCs were backed up from
+
+	for _, pvcRestore := range vmfr.Status.PVCRestores {
+		// Skip synthetic PVC entries (backup-level failures)
+		if pvcRestore.PVCName == "backup-level-failure" {
+			continue
+		}
+
+		// Process each restore info for this PVC
+		for _, restoreInfo := range pvcRestore.Restores {
+			// Only create Velero Restores for available backups
+			if restoreInfo.State != string(oadptypes.BackupDiscoveryStateAvailable) {
+				logger.V(1).Info("Skipping non-available restore",
+					"pvcUID", pvcRestore.PVCUID,
+					"backup", restoreInfo.VeleroBackupName,
+					"state", restoreInfo.State)
+				continue
+			}
+
+			// Add this PVC UID to the backup's list
+			backupKey := restoreInfo.VeleroBackupName
+			if _, exists := backupToPVCUIDs[backupKey]; !exists {
+				backupToPVCUIDs[backupKey] = []string{}
+				// Store the source namespace where the PVCs were backed up from (not the OADP namespace)
+				backupToSourceNamespace[backupKey] = pvcRestore.PVCNamespace
+			}
+			backupToPVCUIDs[backupKey] = append(backupToPVCUIDs[backupKey], pvcRestore.PVCUID)
+		}
+	}
+
+	if len(backupToPVCUIDs) == 0 {
+		return fmt.Errorf("no available backups found to restore - this should have been caught in validation")
+	}
+
+	logger.V(0).Info("Grouped PVCs by backup for restore creation",
+		"backupCount", len(backupToPVCUIDs),
+		"targetNamespace", restoreNamespace)
+
+	// Create one Velero Restore per backup
+	createdRestores := 0
+	for backupName, pvcUIDs := range backupToPVCUIDs {
+		sourceNamespace := backupToSourceNamespace[backupName]
+
+		// Generate Velero Restore name
+		// Format: vmfr-<vmfr-name>-<backup-name>-<8-char-suffix>
+		restoreName := fmt.Sprintf("vmfr-%s-%s-%s",
+			vmfr.Name,
+			backupName,
+			string(vmfr.UID[:8]))
+
+		// Ensure name is DNS-1123 compliant (max 253 chars for Velero Restore)
+		if len(restoreName) > 253 {
+			// Truncate middle part to fit
+			maxMiddleLen := 253 - len("vmfr-") - len(string(vmfr.UID[:8])) - 2 // -2 for hyphens
+			if maxMiddleLen > 0 {
+				middlePart := vmfr.Name + "-" + backupName
+				if len(middlePart) > maxMiddleLen {
+					middlePart = middlePart[:maxMiddleLen]
+				}
+				restoreName = fmt.Sprintf("vmfr-%s-%s", middlePart, string(vmfr.UID[:8]))
+			}
+		}
+		restoreName = strings.ToLower(restoreName)
+
+		logger.V(1).Info("Creating Velero Restore",
+			"restoreName", restoreName,
+			"backupName", backupName,
+			"pvcCount", len(pvcUIDs))
+
+		// Build OrLabelSelectors for all PVC UIDs in this backup
+		labelSelectors := make([]*metav1.LabelSelector, 0, len(pvcUIDs))
+		for _, pvcUID := range pvcUIDs {
+			labelSelectors = append(labelSelectors, &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					constant.PVCUIDLabel: pvcUID,
+				},
+			})
+		}
+
+		// Create Velero Restore object
+		veleroRestore := &veleroapi.Restore{
```

Comment:
Do we need backup name annotation? It's part of the restore spec already, right?

Maybe also add in the annotation VM Name and VM Namespace that is going to be restored from?

```yaml
constant.VirtualMachineNameAnnotation:
constant.VirtualMachineNamespaceAnnotation:
```